xml/ha_migration.xml

```diff
@@ -708,6 +724,111 @@
    </important>
    <para>The &hawk2; <guimenu>Status</guimenu> screen also shows a warning if
     different CRM versions are detected for your cluster nodes.</para>
+   <para>Beside an in-place upgrade, many customers prefer a fresh installation
+    even for moving to the next service pack. The following procedure shows a
+    scenario where a two-node cluster with the nodes &node1; and &node2; is
+    upgraded to the next service pack (SP):
+   </para>
+   <procedure xml:id="pro-ha-migration-rolling-upgrade-newsp-freshinstall">
+    <title>Performing a Cluster-wide Fresh Installation of a New Service Pack</title>
+    <step xml:id="st-ha-migration-rolling-up-sp-backup">
+     <para>Make a backup of your cluster configuration. A minimum set of files
+      are shown in the following list<!--<xref linkend="ex.ha.migration.minlist4backup"
+       xrefstyle="select:label"/>-->:
+     </para>
+     <!--<example xml:id="ex.ha.migration.minlist4backup">
+      <title>Minimum Files for Backup</title>-->
+     <screen>/etc/corosync/corosync.conf
+/etc/corosync/authkey
+/etc/sysconfig/sbd
+/etc/modules-load.d/watchdog.conf
+/etc/hosts
+/etc/ntp.conf</screen>
+     <!--</example>-->
+     <para>Depending on your resources, you may also need the following files:</para>
+     <!--<example xml:id="ex.ha.migration.extlist4backup">
+      <title>Extended List of Files for Backup</title>-->
+     <screen>/etc/services
+/etc/passwd
+/etc/shadow
+/etc/groups
+/etc/drbd/*
+/etc/lvm/lvm.conf
+/etc/mdadm.conf
+/etc/mdadm.SID.conf</screen>
+     <!--</example>-->
+    </step>
+    <step xml:id="st-ha-migration-rolling-alice">
+     <para>Start with node &node1;.
+     </para>
+     <substeps>
+      <step>
+       <para>
+        Make sure no resources are running. If needed,
+        migrate your resource to &node2;. See <xref
+         linkend="sec-conf-hawk2-manage-migrate" xrefstyle="select:label"/>
+        (&hawk;) or <xref
+         linkend="sec-ha-manual-config-migrate" xrefstyle="select:label"/>
+        (&crmshell;)  for details.
```

Comment:
The step makes sense as long as the node has to be rebooted. But not sure how it's specifically related to fresh installation :-) It'd be similar for the case of performing an OS upgrade to the next SP, no?

Reply:
Besides, instead of moving specific resources possibly one by one,  we could put the node into "standby" mode so that all resources on it get moved away:

https://susedoc.github.io/doc-sleha/master/single-html/SLE-HA-guide/#sec-ha-maint-node-standby


BTW, let's use the term "move" rather than "migrating". The latter is confusingly used a lot of also in other occasions... FYI:
https://bugzilla.suse.com/show_bug.cgi?id=1166644

Reply:
@gao-yan: I like the idea of putting a cluster node into standby mode instead of moving a resource one by one.

However, I would clarify the overall structure of the procedure with you. So my (uneducated) guess of the overall procedure would be:

* Step 1: "Make a backup of your cluster configuration"
    I guess, that can stay. :)
* Step 2: "Start with node alice."

    * Step a: "Put the node into standby node to move all resources to the other node."
         This would be the modified step. Instead of moving the resources, we put the node itself into standby mode.
    * Step b: "Check that the cluster in healthy and idle."
       I changed it slightly and added the command "crm cluster health". Ok?
    * Step c: ~"Set both nodes in maintenance mode"~
     This step can be removed as mentioned in your later comment.
    * Step d: "Stop Pacemaker on alice"
      Should stay, this time with "crm cluster stop"
    * Step e: "Upgrade empty nodes and reboot."
     Rephrased according to my comment https://github.com/SUSE/doc-sleha/pull/130#discussion_r408615008
    * Step f: "Copy your backup files..."
     Should stay of course.
   * Step g: "Bring back ~empty~ node alice ..."
    Should stay too, but without "empty".
   * Step h: "Check if alice is fine."
    Hmn, how? Just with "crm status"? Or again with "crm cluster health"? Or remove the step completely?
   * Step i: "Move all resources to the upgrade node alice"
     Is this still needed? I guess, the "standby" mode isn't preserved after a system upgrade. So does we still need to thell the cluster or is it done automatically?
* Step 3: "Repeat..."
  That needs to stay of course.

Thanks!

Reply:
Here is a little picture of the above steps to get an impression:

----

![upgrade](https://user-images.githubusercontent.com/1312925/79309927-5d929500-7efb-11ea-8407-3e56155e7edb.png)

Reply:
> * Step 1: "Make a backup of your cluster configuration"
>   I guess, that can stay. :)

For sure :-)

> * Step 2: "Start with node alice."
>   
>   * Step a: "Put the node into standby node to move all resources to the other node."
>     This would be the modified step. Instead of moving the resources, we put the node itself into standby mode.
>   * Step b: "Check that the cluster in healthy and idle."
>     I changed it slightly and added the command "crm cluster health". Ok?

We actually could combine these two steps with this command:
```
crm -w node standby alice
```
With `-w`, the command waits for the cluster to finish the transition and become idle.

I'm actually thinking it'd probably be even better to add a `reboot` option for it:
```
crm -w node standby alice reboot
```
With the `reboot` option, `standby` mode applies only until cluster services on `alice` are stopped. So that when `alice` is online again, we won't need to run a command `crm node online alice` to take it out from `standby` mode.

This command is used also in here:
https://susedoc.github.io/doc-sleha/master/single-html/SLE-HA-guide/#pro-ha-maint-shutdown-node

We should probably use the same way there.

-- Well, the term `migrate` used there should be changed to `move` as well :-)

Besides, do you think this should be a commonly recommended step which could be applied also for the normal rolling upgrade:

https://susedoc.github.io/doc-sleha/master/single-html/SLE-HA-guide/#pro-ha-migration-rolling-upgrade

?

>   * Step c: ~"Set both nodes in maintenance mode"~
>     This step can be removed as mentioned in your later comment.

Agreed.

>   * Step d: "Stop Pacemaker on alice"

Let's say "Stop cluster services" rather than "Stop Pacemaker", and correspondingly "Start cluster services" rather than "Start Pacemaker" if there's any. I know for historical reason, these are used everywhere in our HA docs. It's probably a good opportunity to change all of them now :-)

>     Should stay, this time with "crm cluster stop"
>   * Step e: "Upgrade empty nodes and reboot."
>     Rephrased according to my comment [#130 (comment)](https://github.com/SUSE/doc-sleha/pull/130#discussion_r408615008)
>   * Step f: "Copy your backup files..."
>     Should stay of course.
>   * Step g: "Bring back ~empty~ node alice ..."
>     Should stay too, but without "empty".
>   * Step h: "Check if alice is fine."
>     Hmn, how? Just with "crm status"? Or again with "crm cluster health"? Or remove the step completely?

Let's probably remove this. Users will somehow figure that out themselves :-)

>   * Step i: "Move all resources to the upgrade node alice"
>     Is this still needed? I guess, the "standby" mode isn't preserved after a system upgrade. So does we still need to thell the cluster or is it done automatically?

As mentioned, with "reboot" option used in above, we could remove this step as well.

> * Step 3: "Repeat..."
>   That needs to stay of course.

It generally looks good to me. Good summary, Thomas. Thank!

Reply:
Thanks Yan! :+1: 

> We actually could combine these two steps with this command:

Even better! :+1:  Ok, so I've combined these:

```xml
<step>
    <para>
        Put the node into standby node to move all resources to the
        other node:
    </para>
    <screen>&prompt.root;<command>crm</command> --wait node standby &node1; reboot</screen>
    <para>
        With the option <option>--wait</option> the command waits for the cluster
        to finish the transition and becomes idle.
        With the <option>reboot</option> option you specify a lifetime for the standby.
        This has the effect that the node will be back online once it reboots.
    </para>
</step>
```

I've used the long option (`--wait` instead of `-w`) as this is a bit more descriptive.

However, should I still add the "crm cluster health" command after "crm -W node standby alice reboot"? Would it still be necessary?

---

> Besides, do you think this should be a commonly recommended step which could be applied also for the normal rolling upgrade:

Of course! :+1: That would improve consistency and avoids confusing our customers.

---
> Let's say "Stop cluster services" rather than "Stop Pacemaker", and correspondingly "Start cluster services" rather than "Start Pacemaker" if there's any

Yes, absolutely agree! :+1: I've opened a separate Trello Card: https://trello.com/c/YGxydKyb

---

I've removed the unnecessary steps as advised.

---

> It generally looks good to me. Good summary, Thomas. Thank!

Thanks due to **your** help. :+1: 

I will post an updated picture so it may be a bit easier to read.

Reply:
![upgrade-2](https://user-images.githubusercontent.com/1312925/79322014-853e2900-7f0c-11ea-9969-6bdc77c5bf30.png)

Reply:
> 
> ```
> <step>
>     <para>
>         Put the node into standby node to move all resources to the
>         other node:
>     </para>
>     <screen>&prompt.root;<command>crm</command> --wait node standby &node1; reboot</screen>
>     <para>
>         With the option <option>--wait</option> the command waits for the cluster
>         to finish the transition and becomes idle.

It's the cluster to become idle rather than the command :-)  Not sure how it'd be clearer. Could be something like "With the option `--wait`, the command returns only when the cluster finishes the transition and becomes idle." Anyway you must have better ideas on the phrasing :-)


>         With the <option>reboot</option> option you specify a lifetime for the standby.
>         This has the effect that the node will be back online once it reboots.

I don't really like the naming of the sub-command `online`, nor the option `reboot`. They are confusing, but somehow were chosen and have to remain for compatibility ... Anyway we could make the description clear. Something like "This has the effect that the node will be already out of standby mode when it is online again".

>     </para>
> </step>
> ```
> 
> I've used the long option (`--wait` instead of `-w`) as this is a bit more descriptive.

Nice idea👍 

> 
> However, should I still add the "crm cluster health" command after "crm -W node standby alice reboot"? Would it still be necessary?

Not really necessary for the procedure. Although users would likely check the status somehow at this point. But that could be done their preferred ways based on their experience.

Reply:
> >  With the option <option>--wait</option> the command waits for the cluster
> >  to finish the transition and becomes idle.
> It's the cluster to become idle rather than the command :-) 

You're absolutely right. :smile: :+1: 

> Not sure how it'd be clearer. Could be something like "With the option `--wait`, the command returns only when the cluster finishes the transition and becomes idle." Anyway you must have better ideas on the phrasing :-)

I don't have a better phrase, I think it's fine now.

 
> [...]
> I don't really like the naming of the sub-command `online`, nor the option `reboot`. They are confusing, but somehow were chosen and have to remain for compatibility ... Anyway we could make the description clear. Something like "This has the effect that the node will be already out of standby mode when it is online again".

I think the "out of standby mode" is a bit foggy. :) From my understanding, it basically means, the cluster node is online again (without being in standby mode), right?

So what do you think about this phrase:

```
The <option>reboot</option> option has the effect that the node will
become online again after a reboot. Any standby status has been reset.
```


> [...]
> > However, should I still add the "crm cluster health" command after "crm -W node standby alice reboot"? Would it still be necessary?
> 
> Not really necessary for the procedure. Although users would likely check the status somehow at this point. But that could be done their preferred ways based on their experience.

Ok, so I leave it better out.

Seems we are almost through. :wink:

Reply:
> > I don't really like the naming of the sub-command `online`, nor the option `reboot`. They are confusing, but somehow were chosen and have to remain for compatibility ... Anyway we could make the description clear. Something like "This has the effect that the node will be already out of standby mode when it is online again".
> 
> I think the "out of standby mode" is a bit foggy. :) From my understanding, it basically means, the cluster node is online again (without being in standby mode), right?

Not really, depending on how you define "online". Technically cluster is online as long as its part of membership. So a node could be "online" in that terms, but meanwhile is "standby" in terms of the role regarding hosting resources. So a node could be "online",  "online standby", or "online maintenance" and so on.

That's exactly the reason why  it's such a big confusion that sub-command name "online" was chosen for taking a node out of standby mode ... In short, technically "online" is not the opposite of "standby".

> 
> So what do you think about this phrase:
> 
> ```
> The <option>reboot</option> option has the effect that the node will
> become online again after a reboot. Any standby status has been reset.
> ```

And again,  the bad naming of the option"reboot". A node doesn't have to "reboot" to get this.  It works as long as a node goes offline and online again.

Reply:
Apparently the help text of crmsh itself is confusing about this, which needs to be fixed.

Reply:
> [...]
Ahh, ok, I see what you mean. My naive understanding was indeed "online" is the opposite of "standby". Thanks for helping me to understand this. :+1: 

Ok, you convinced me. :smile: In that case, we have to mention "standby" in one form or the other. For the future, maybe we could create a term which would mean the opposite. 

Just out of curiosity, the opposite of "standby" is a running node, which is part of the cluster and not fenced... anything else? Phuu, I guess, that is hard to come up with a single word. :wink: 

> And again, the bad naming of the option"reboot". A node doesn't have to "reboot" to get this. It works as long as a node goes offline and online again.

Ok, next attempt:

```
The <option>reboot</option> option has the effect that the node will reset
its standby mode when it is online again.
```

Maybe we could add a short hint about the (bad) naming:

```
Despite the name, the reboot option works also when the node goes offline and online.
```

Reply:
> Apparently the help text of crmsh itself is confusing about this, which needs to be fixed.

It is. It is awfully close to an old Douglas Adams joke: :smile: 

> "Ford flipped the switch which he saw was marked "Mode Execute Ready" instead of the now old-fashioned "Access Standby" that had so long ago replaced the appallingly stone-aged "Off."

Reply:
I consulted our native language experts and explained the background. 

So the opposite of "standby mode" could be:

* "active" (this reminds me of active/active configuration)
* "live" (would that work? Not sure...)

Reply:
> > Ahh, ok, I see what you mean. My naive understanding was indeed "online" is the opposite of "standby". Thanks for helping me to understand this. 👍
> 
> Ok, you convinced me. 😄 In that case, we have to mention "standby" in one form or the other. For the future, maybe we could create a term which would mean the opposite.
> 
> Just out of curiosity, the opposite of "standby" is a running node, which is part of the cluster and not fenced... anything else? Phuu, I guess, that is hard to come up with a single word. 😉

Non-standby? :-)  Anyway, it's just one of the properties of a node. For example, another property is "maintenance". So a node could be even "online standby maintenance". We could never invent enough single-words for all the combinations :-).  IIRC, description like "out of maintenance mode" is used somewhere ...

> 
> > And again, the bad naming of the option"reboot". A node doesn't have to "reboot" to get this. It works as long as a node goes offline and online again.
> 
> Ok, next attempt:
> 
> ```
> The <option>reboot</option> option has the effect that the node will reset
> its standby mode when it is online again.
> ```

Hrm, not really clear to me that reset means to set it "on" or "off" :-)

> 
> Maybe we could add a short hint about the (bad) naming:
> 
> ```
> Despite its name, the reboot option works also when the node goes offline and online.
> ```

Or s/also when/as long as/ ? :-)

Reply:
> > Just out of curiosity, the opposite of "standby" is a running node, which is part of the cluster and not fenced... anything else? Phuu, I guess, that is hard to come up with a single word. wink
> 
> Non-standby? :-) Anyway, it's just one of the properties of a node. For example, another property is "maintenance". So a node could be even "online standby maintenance". We could never invent enough single-words for all the combinations :-). IIRC, description like "out of maintenance mode" is used somewhere ...

Ahh, you define it in reverse logic! Clever! :+1:  I think that's the closest we can get. So I like "non-standby". :smile: 


> > Ok, next attempt:
> > ```
> > The <option>reboot</option> option has the effect that the node will reset
> > its standby mode when it is online again.
> > ```
> 
> Hrm, not really clear to me that reset means to set it "on" or "off" :-)

You are absolutely right. Argh. My language skills are getting tired. :laughing:  Ok, I will use the one you came up first:

```
The <option>reboot</option> option has the effect that the node will
be already out of standby mode when it is online again.
```

Maybe I need to rethink this with a fresh brain tomorrow.

> Or s/also when/as long as/ ? :-)

Yes, fixed that.